src/main/java/backend/RepoIO.java

```diff
@@ -37,20 +34,19 @@
 
     private static final int MAX_REDOWNLOAD_TRIES = 2;
 
-    public RepoIO(boolean isTestMode, boolean enableTestJSON) {
-        if (isTestMode) {
-            repoSource = new DummySource();
-            RepoStore.enableTestDirectory();
-        } else {
-            repoSource = new GitHubSource();
-        }
-        if (isTestMode && !enableTestJSON) {
-            jsonStore = new JSONStoreStub();
-            storedRepos = new ArrayList<>();
-        } else {
-            jsonStore = new JSONStore();
-            storedRepos = new ArrayList<>(jsonStore.getStoredRepos());
-        }
+    /**
+     * Contructs a RepoIO providing IO operations on repositories, taking in various optional
+     * parameters for repos source and storage which are useful for testing purposes.
+     * @param repoSource optional source of repos. Default to GitHubSource if not present
+     * @param jsonStore optional storage for repos. Default to a new JSONStore if not present
+     * @param storeDirectory optional directory for storing repos. Default value is in RepoStore.
+     */
+    public RepoIO(Optional<RepoSource> repoSource, Optional<JSONStore> jsonStore,
+                  Optional<String> storeDirectory) {
+        this.repoSource = repoSource.orElseGet(() -> new GitHubSource());
+        storeDirectory.ifPresent((dir) -> RepoStore.changeDirectory(dir));
+        this.jsonStore = jsonStore.orElseGet(() -> new JSONStore());
```

Comment:
The first line is to avoid creating a new instance of `GitHubSource` (which takes quite some time) when it's not needed. I will change to method's reference in the second line.